internal/luks2/luks2.go

```diff
@@ -0,0 +1,968 @@
+			return xerrors.Errorf("invalid keyslot index: %w", err)
+		}
+		m.Keyslots[id] = v
+	}
+
+	m.Segments = make(map[int]*Segment)
+	for k, v := range d.Segments {
+		id, err := k.int()
+		if err != nil {
+			return xerrors.Errorf("invalid segment index: %w", err)
+		}
+		m.Segments[id] = v
+	}
+
+	m.Digests = make(map[int]*Digest)
+	for k, v := range d.Digests {
+		id, err := k.int()
+		if err != nil {
+			return xerrors.Errorf("invalid digest index: %w", err)
+		}
+		m.Digests[id] = v
+	}
+
+	m.Tokens = make(map[int]*Token)
+	for k, v := range d.Tokens {
+		id, err := k.int()
+		if err != nil {
+			return xerrors.Errorf("invalid token index: %w", err)
+		}
+		m.Tokens[id] = v
+	}
+
+	m.Config = d.Config
+	return nil
+}
+
+type HdrInfo struct {
+	HdrSize  uint64
+	Label    string
+	Metadata Metadata
+}
+
+func toString(b []byte) string {
+	return strings.TrimRight(string(b), "\x00")
+}
+
+func getHash(alg string) crypto.Hash {
+	switch alg {
+	case "sha1":
+		return crypto.SHA1
+	case "sha224":
+		return crypto.SHA224
+	case "sha256":
+		return crypto.SHA256
+	case "sha384":
+		return crypto.SHA384
+	case "sha512":
+		return crypto.SHA512
+	default:
+		return 0
+	}
+}
+
+func decodeAndCheckHeader(r io.ReadSeeker, offset int64, primary bool) (*binaryHdr, []byte, error) {
+	if _, err := r.Seek(offset, io.SeekStart); err != nil {
+		return nil, nil, err
+	}
+
+	var hdr binaryHdr
+	if err := binary.Read(r, binary.BigEndian, &hdr); err != nil {
+		return nil, nil, xerrors.Errorf("cannot read header: %w", err)
+	}
+	switch {
+	case primary && bytes.Equal(hdr.Magic[:], []byte("LUKS\xba\xbe")):
+	case !primary && bytes.Equal(hdr.Magic[:], []byte("SKUL\xba\xbe")):
+	default:
+		return nil, nil, errors.New("invalid magic")
+	}
+	if hdr.Version != 2 {
+		return nil, nil, errors.New("invalid version")
+	}
+	if hdr.HdrSize > uint64(math.MaxInt64) {
+		return nil, nil, errors.New("header size too large")
+	}
+	if hdr.HdrOffset > uint64(math.MaxInt64) {
+		return nil, nil, errors.New("header offset too large")
+	}
+	if int64(hdr.HdrOffset) != offset {
+		return nil, nil, errors.New("invalid header offset")
+	}
+
+	csumHash := getHash(toString(hdr.CsumAlg[:]))
+	if csumHash == 0 {
+		return nil, nil, errors.New("unsupported checksum alg")
+	}
+
+	h := csumHash.New()
+
+	hdrTmp := hdr
+	hdrTmp.Csum = [64]byte{}
+
+	if err := binary.Write(h, binary.BigEndian, &hdrTmp); err != nil {
+		return nil, nil, xerrors.Errorf("cannot calculate checksum, error serializing header: %w", err)
+	}
+
+	var jsonBuffer bytes.Buffer
+	tr := io.TeeReader(r, &jsonBuffer)
+
+	if _, err := io.CopyN(h, tr, int64(hdr.HdrSize)-int64(binary.Size(hdr))); err != nil {
+		return nil, nil, xerrors.Errorf("cannot calculate checksum, error reading JSON metadata: %w", err)
+	}
+
+	if !bytes.Equal(h.Sum(nil), hdr.Csum[0:csumHash.Size()]) {
+		return nil, nil, errors.New("invalid header checksum")
+	}
+
+	return &hdr, jsonBuffer.Bytes(), nil
+}
+
+func DecodeHdr(path string) (*HdrInfo, error) {
```

Comment:
To make this landable, I think this would need to be submitted first with its own tests and with pointers to upstream code/docs as its own PR. Then the rest of luks2 as its own PR, then finally its usage from crypt.go etc.